Shepherd Judge phase fails silently with 0s session duration

## Problem

The shepherd's Judge phase can fail silently when the spawned Claude session produces zero output. The session starts but immediately exits with no approval comment, no rejection comment, and no label changes — just `session duration: 0s`.

The shepherd retries once (as configured) and gets the same result, then fails with exit code 4.

## Observed Behavior

```
[force-mode] Label validation failed for PR #2131, attempting fallback approval detection
[force-mode] No approval comment found in PR — fallback denied
[force-mode] Approval fallback failed for PR #2131, attempting fallback changes-requested detection
[force-mode] No rejection comment or CHANGES_REQUESTED review found — fallback denied
PR #2131 has loom:review-requested (Doctor applied fixes) but judge did not produce outcome label
[ERROR] Judge phase failed after 1 retry attempt(s): judge phase validation failed:
  no GitHub reviews on PR (Loom uses comment + label workflow instead);
  labels=[loom:review-requested, loom:reviewing];
  last output: '';
  session duration: 0s
```

## Key Indicators

- `session duration: 0s` — Claude session started but produced no output
- `last output: ''` — Empty output from the Judge agent
- Labels stuck at `[loom:review-requested, loom:reviewing]` — Judge claimed the PR but never completed
- Retry produced the same 0s result

## Impact

- Shepherd exits with code 4 (Judge failure), leaving PR in limbo
- PR has `loom:reviewing` label but no review was actually performed
- Requires manual intervention to evaluate and merge the PR
- In daemon mode, this would leave a shepherd slot occupied with a failed issue

## Context

- Occurred during `/shepherd 2128 -m` on 2026-02-05
- PR #2131 was CLEAN merge state, CI passing
- Manual Judge evaluation (by human) approved the PR immediately — the code was fine
- The issue appears to be with the Claude session startup, not the PR content

## Possible Causes

1. Claude CLI session failed to initialize (transient API error)
2. Context window issue with the Judge role prompt
3. Race condition in session spawning
4. Agent crashed before producing any tool calls

## Suggested Investigation

- Check if `agent-spawn.sh` or `agent-wait.sh` logs capture the failed session
- Consider adding a minimum session duration check (if <5s, likely a startup failure)
- Consider increasing retry count for 0s sessions specifically (transient failures)
- Add diagnostic logging when session duration is suspiciously short

--- a/loom-tools/src/loom_tools/shepherd/phases/base.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/base.py
@@ -11,6 +11,8 @@ from enum import Enum
 from pathlib import Path
 from typing import TYPE_CHECKING, Any, Protocol
 
+from loom_tools.common.logging import log_warning, strip_ansi
+from loom_tools.common.paths import LoomPaths
 from loom_tools.common.state import read_json_file
 
 if TYPE_CHECKING:
@@ -18,6 +20,20 @@ if TYPE_CHECKING:
 
 # How often (in seconds) to poll the progress file during agent-wait-bg.sh
 _HEARTBEAT_POLL_INTERVAL = 5
+
+# Minimum session duration (seconds) to consider a session as having run
+# successfully. Sessions shorter than this with no meaningful output are
+# treated as transient spawn failures (e.g., Claude API error on startup).
+# See issue #2135.
+INSTANT_EXIT_THRESHOLD_SECONDS = 5
+
+# Minimum characters of non-ANSI content required for "meaningful output".
+# Matches the threshold used in judge.py for consistency.
+INSTANT_EXIT_MIN_OUTPUT_CHARS = 100
+
+# Maximum retries for instant-exit detection, with exponential backoff.
+INSTANT_EXIT_MAX_RETRIES = 3
+INSTANT_EXIT_BACKOFF_SECONDS = [2, 4, 8]
 
 
 class PhaseStatus(Enum):
@@ -258,6 +274,39 @@ def _print_heartbeat(action: str) -> None:
     print(f"\033[2m[{ts}] \u27f3 {action}\033[0m", file=sys.stderr)
 
 
+def _is_instant_exit(log_path: Path) -> bool:
+    """Check if a session log indicates an instant-exit (transient spawn failure).
+
+    A session is considered an instant exit when:
+    - The log file exists but is very short-lived (< INSTANT_EXIT_THRESHOLD_SECONDS)
+    - AND has no meaningful output (< INSTANT_EXIT_MIN_OUTPUT_CHARS non-ANSI chars)
+
+    This detects cases where the Claude CLI spawns but immediately exits due to
+    transient API errors, without producing any substantive work.
+
+    Args:
+        log_path: Path to the worker session log file.
+
+    Returns:
+        True if the session appears to be an instant exit.
+    """
+    if not log_path.is_file():
+        # No log file at all — could be spawn failure, not instant exit.
+        return False
+
+    try:
+        stat = log_path.stat()
+        duration = int(stat.st_mtime - stat.st_ctime)
+        if duration >= INSTANT_EXIT_THRESHOLD_SECONDS:
+            return False
+
+        content = log_path.read_text()
+        stripped = strip_ansi(content)
+        return len(stripped.strip()) < INSTANT_EXIT_MIN_OUTPUT_CHARS
+    except OSError:
+        return False
+
+
 def run_worker_phase(
     ctx: ShepherdContext,
     *,
@@ -286,11 +335,12 @@ def run_worker_phase(
         args: Optional arguments for the worker
 
     Returns:
-        Exit code from agent-wait-bg.sh:
+        Exit code from agent-wait-bg.sh (or synthetic):
         - 0: Success
         - 3: Shutdown signal
         - 4: Agent stuck after retry
         - 5: Failures are pre-existing (Doctor only)
+        - 6: Instant exit detected (session < 5s with no meaningful output)
         - Other: Error
     """
     scripts_dir = ctx.scripts_dir
@@ -396,6 +446,20 @@ def run_worker_phase(
         check=False,
     )
 
+    # Detect instant-exit: session completed (exit 0) but ran for < 5s with
+    # no meaningful output.  Return synthetic exit code 6 so the retry layer
+    # can handle it with backoff.  See issue #2135.
+    if wait_exit == 0:
+        paths = LoomPaths(ctx.repo_root)
+        log_path = paths.worker_log_file(role, ctx.config.issue)
+        if _is_instant_exit(log_path):
+            log_warning(
+                f"Instant-exit detected for {role} session '{name}': "
+                f"session completed in <{INSTANT_EXIT_THRESHOLD_SECONDS}s "
+                f"with no meaningful output (log: {log_path})"
+            )
+            return 6
+
     return wait_exit
 
 
@@ -411,14 +475,18 @@ def run_phase_with_retry(
     pr_number: int | None = None,
     args: str | None = None,
 ) -> int:
-    """Run a phase with automatic retry on stuck detection.
+    """Run a phase with automatic retry on stuck or instant-exit detection.
 
     On exit code 4 (stuck), retries up to max_retries times.
+    On exit code 6 (instant exit), retries up to INSTANT_EXIT_MAX_RETRIES
+    times with exponential backoff.
 
     Returns:
-        Exit code: 0=success, 3=shutdown, 4=stuck after retries, other=error
+        Exit code: 0=success, 3=shutdown, 4=stuck after retries,
+                   6=instant-exit after retries, other=error
     """
     stuck_retries = 0
+    instant_exit_retries = 0
 
     while True:
         exit_code = run_worker_phase(
@@ -432,8 +500,41 @@ def run_phase_with_retry(
             args=args,
         )
 
+        # --- Instant-exit handling (exit code 6) ---
+        if exit_code == 6:
+            instant_exit_retries += 1
+            if instant_exit_retries > INSTANT_EXIT_MAX_RETRIES:
+                log_warning(
+                    f"Instant-exit persisted for {role} after "
+                    f"{INSTANT_EXIT_MAX_RETRIES} retries"
+                )
+                return 6  # Caller should treat as failure
+
+            # Exponential backoff before retry
+            backoff_idx = min(
+                instant_exit_retries - 1, len(INSTANT_EXIT_BACKOFF_SECONDS) - 1
+            )
+            backoff = INSTANT_EXIT_BACKOFF_SECONDS[backoff_idx]
+
+            ctx.report_milestone(
+                "error",
+                error=f"instant-exit detected for {role}",
+                will_retry=True,
+            )
+            ctx.report_milestone(
+                "heartbeat",
+                action=(
+                    f"retrying instant-exit {role} "
+                    f"(attempt {instant_exit_retries}/{INSTANT_EXIT_MAX_RETRIES}, "
+                    f"backoff {backoff}s)"
+                ),
+            )
+
+            time.sleep(backoff)
+            continue
+
+        # --- Stuck handling (exit code 4) ---
         if exit_code != 4:
-            # Not stuck - return as-is
             return exit_code
 
         stuck_retries += 1

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -24,8 +24,14 @@ from loom_tools.shepherd.phases import (
     PhaseStatus,
 )
 from loom_tools.shepherd.phases.base import (
+    INSTANT_EXIT_BACKOFF_SECONDS,
+    INSTANT_EXIT_MAX_RETRIES,
+    INSTANT_EXIT_MIN_OUTPUT_CHARS,
+    INSTANT_EXIT_THRESHOLD_SECONDS,
+    _is_instant_exit,
     _print_heartbeat,
     _read_heartbeats,
+    run_phase_with_retry,
     run_worker_phase,
 )
 from loom_tools.shepherd.phases.judge import (
@@ -8816,3 +8822,360 @@ class TestDoctorCIWaiting:
         assert result["checks_passed"] == 2
         assert result["checks_failed"] == 0
         assert result["checks_pending"] == 1
+
+
+# ---------------------------------------------------------------------------
+# Instant-exit detection tests (issue #2135)
+# ---------------------------------------------------------------------------
+
+
+class TestIsInstantExit:
+    """Test _is_instant_exit helper function."""
+
+    def test_no_log_file_returns_false(self, tmp_path: Path) -> None:
+        """Missing log file should not be flagged as instant exit."""
+        assert _is_instant_exit(tmp_path / "nonexistent.log") is False
+
+    def test_short_session_no_output_returns_true(self, tmp_path: Path) -> None:
+        """Log with <5s duration and minimal content is an instant exit."""
+        log = tmp_path / "session.log"
+        # Write only ANSI escape sequences (no meaningful content)
+        log.write_text("\x1b[?2026l\x1b[0m\n")
+        # File ctime and mtime will be nearly identical (0s apart)
+        assert _is_instant_exit(log) is True
+
+    def test_short_session_with_output_returns_false(self, tmp_path: Path) -> None:
+        """Log with <5s duration but meaningful content is NOT an instant exit."""
+        log = tmp_path / "session.log"
+        # Write enough meaningful content to exceed threshold
+        log.write_text("x" * (INSTANT_EXIT_MIN_OUTPUT_CHARS + 1))
+        assert _is_instant_exit(log) is False
+
+    def test_long_session_no_output_returns_false(self, tmp_path: Path) -> None:
+        """Log with >=5s duration but empty content is NOT an instant exit.
+
+        This tests that the duration check happens first (short-circuit).
+        """
+        import os
+
+        log = tmp_path / "session.log"
+        log.write_text("\x1b[?2026l")
+        # Fake a 10-second session by backdating the ctime via atime/mtime
+        stat = log.stat()
+        os.utime(log, (stat.st_atime, stat.st_mtime + 10))
+        assert _is_instant_exit(log) is False
+
+    def test_empty_log_returns_true(self, tmp_path: Path) -> None:
+        """Completely empty log with 0s duration is an instant exit."""
+        log = tmp_path / "session.log"
+        log.write_text("")
+        assert _is_instant_exit(log) is True
+
+
+class TestRunWorkerPhaseInstantExit:
+    """Test that run_worker_phase detects instant exits and returns code 6."""
+
+    @pytest.fixture
+    def mock_context(self) -> MagicMock:
+        """Create a mock ShepherdContext."""
+        ctx = MagicMock(spec=ShepherdContext)
+        ctx.config = ShepherdConfig(issue=42, task_id="test-123")
+        ctx.repo_root = Path("/fake/repo")
+        ctx.scripts_dir = Path("/fake/repo/.loom/scripts")
+        ctx.progress_dir = Path("/tmp/progress")
+        return ctx
+
+    def test_instant_exit_returns_code_6(self, mock_context: MagicMock) -> None:
+        """When agent completes in <5s with no output, return exit code 6."""
+
+        def mock_spawn(cmd, **kwargs):
+            result = MagicMock()
+            result.returncode = 0
+            return result
+
+        def mock_popen(cmd, **kwargs):
+            proc = MagicMock()
+            proc.poll.return_value = 0
+            proc.returncode = 0
+            return proc
+
+        with (
+            patch("subprocess.run", side_effect=mock_spawn),
+            patch("subprocess.Popen", side_effect=mock_popen),
+            patch("time.sleep"),
+            patch(
+                "loom_tools.shepherd.phases.base._is_instant_exit", return_value=True
+            ),
+        ):
+            exit_code = run_worker_phase(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                phase="judge",
+            )
+
+        assert exit_code == 6
+
+    def test_normal_completion_returns_code_0(self, mock_context: MagicMock) -> None:
+        """When agent completes normally, return exit code 0."""
+
+        def mock_spawn(cmd, **kwargs):
+            result = MagicMock()
+            result.returncode = 0
+            return result
+
+        def mock_popen(cmd, **kwargs):
+            proc = MagicMock()
+            proc.poll.return_value = 0
+            proc.returncode = 0
+            return proc
+
+        with (
+            patch("subprocess.run", side_effect=mock_spawn),
+            patch("subprocess.Popen", side_effect=mock_popen),
+            patch("time.sleep"),
+            patch(
+                "loom_tools.shepherd.phases.base._is_instant_exit", return_value=False
+            ),
+        ):
+            exit_code = run_worker_phase(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                phase="judge",
+            )
+
+        assert exit_code == 0
+
+    def test_non_zero_exit_skips_instant_exit_check(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Exit codes other than 0 should not trigger instant-exit detection."""
+
+        def mock_spawn(cmd, **kwargs):
+            result = MagicMock()
+            result.returncode = 0
+            return result
+
+        def mock_popen(cmd, **kwargs):
+            proc = MagicMock()
+            proc.poll.return_value = 0
+            proc.returncode = 4  # Stuck
+            return proc
+
+        with (
+            patch("subprocess.run", side_effect=mock_spawn),
+            patch("subprocess.Popen", side_effect=mock_popen),
+            patch("time.sleep"),
+            patch(
+                "loom_tools.shepherd.phases.base._is_instant_exit"
+            ) as mock_check,
+        ):
+            exit_code = run_worker_phase(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                phase="judge",
+            )
+
+        assert exit_code == 4
+        mock_check.assert_not_called()
+
+
+class TestRunPhaseWithRetryInstantExit:
+    """Test that run_phase_with_retry retries on instant-exit (code 6)."""
+
+    @pytest.fixture
+    def mock_context(self) -> MagicMock:
+        """Create a mock ShepherdContext."""
+        ctx = MagicMock(spec=ShepherdContext)
+        ctx.config = ShepherdConfig(
+            issue=42, task_id="test-123", stuck_max_retries=2
+        )
+        ctx.repo_root = Path("/fake/repo")
+        ctx.scripts_dir = Path("/fake/repo/.loom/scripts")
+        ctx.progress_dir = Path("/tmp/progress")
+        return ctx
+
+    def test_retries_on_instant_exit_then_succeeds(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Should retry on code 6 and return 0 when the retry succeeds."""
+        call_count = 0
+
+        def mock_run_worker(*args, **kwargs):
+            nonlocal call_count
+            call_count += 1
+            # First call: instant exit; second call: success
+            return 6 if call_count == 1 else 0
+
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                side_effect=mock_run_worker,
+            ),
+            patch("time.sleep") as mock_sleep,
+        ):
+            exit_code = run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        assert exit_code == 0
+        assert call_count == 2
+        # First backoff is 2 seconds
+        mock_sleep.assert_called_once_with(INSTANT_EXIT_BACKOFF_SECONDS[0])
+
+    def test_exhausts_retries_returns_code_6(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Should return 6 after exhausting all instant-exit retries."""
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                return_value=6,
+            ),
+            patch("time.sleep"),
+        ):
+            exit_code = run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        assert exit_code == 6
+
+    def test_retry_count_matches_constant(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Should retry exactly INSTANT_EXIT_MAX_RETRIES times."""
+        call_count = 0
+
+        def count_calls(*args, **kwargs):
+            nonlocal call_count
+            call_count += 1
+            return 6
+
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                side_effect=count_calls,
+            ),
+            patch("time.sleep"),
+        ):
+            run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        # 1 initial call + INSTANT_EXIT_MAX_RETRIES retries
+        assert call_count == 1 + INSTANT_EXIT_MAX_RETRIES
+
+    def test_backoff_timing(self, mock_context: MagicMock) -> None:
+        """Should use exponential backoff from INSTANT_EXIT_BACKOFF_SECONDS."""
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                return_value=6,
+            ),
+            patch("time.sleep") as mock_sleep,
+        ):
+            run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        # Verify backoff values for each retry
+        sleep_values = [call.args[0] for call in mock_sleep.call_args_list]
+        assert sleep_values == INSTANT_EXIT_BACKOFF_SECONDS
+
+    def test_reports_error_and_heartbeat_milestones(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Should report error milestone with will_retry and heartbeat milestone."""
+        call_count = 0
+
+        def mock_run_worker(*args, **kwargs):
+            nonlocal call_count
+            call_count += 1
+            return 6 if call_count == 1 else 0
+
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                side_effect=mock_run_worker,
+            ),
+            patch("time.sleep"),
+        ):
+            run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        # Check that milestones were reported
+        milestone_calls = mock_context.report_milestone.call_args_list
+        # Should have: error milestone, then heartbeat milestone
+        assert len(milestone_calls) == 2
+        # First call: error with will_retry
+        assert milestone_calls[0].args[0] == "error"
+        assert milestone_calls[0].kwargs["will_retry"] is True
+        # Second call: heartbeat with retry info
+        assert milestone_calls[1].args[0] == "heartbeat"
+        assert "instant-exit" in milestone_calls[1].kwargs["action"]
+
+    def test_stuck_and_instant_exit_have_separate_counters(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Stuck retries (code 4) and instant-exit retries (code 6) are independent."""
+        call_count = 0
+
+        def mock_run_worker(*args, **kwargs):
+            nonlocal call_count
+            call_count += 1
+            if call_count == 1:
+                return 6  # Instant exit
+            elif call_count == 2:
+                return 4  # Stuck
+            else:
+                return 0  # Success
+
+        with (
+            patch(
+                "loom_tools.shepherd.phases.base.run_worker_phase",
+                side_effect=mock_run_worker,
+            ),
+            patch("time.sleep"),
+        ):
+            exit_code = run_phase_with_retry(
+                mock_context,
+                role="judge",
+                name="judge-issue-42",
+                timeout=600,
+                max_retries=2,
+                phase="judge",
+            )
+
+        assert exit_code == 0
+        assert call_count == 3

